hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/impl/SqlClientConnection.java

```diff
@@ -199,45 +204,87 @@ private void feedback(String sql) {
+	private static final boolean RUNNING_ON_VERTX3 = BEGIN_ON_3 != null;
+
+	private static Method identifyCloseMethodOnVertxVersion3() {
+		try {
+			final Method method = SqlClient.class.getMethod( "close" );
+			final Class<?> returnType = method.getReturnType();
+			// if it returns void
+			return returnType.equals( Void.TYPE )
+					// then we're on Vert.x v. 3.9
+					? method
+					// else it's Vert.x 4
+					: null;
+		}
+		catch (NoSuchMethodException e) {
+			return null; //some new Vert.x version? No need for Vertx 3 compatibility in this case.
+		}
+	}
+
+	private static Method identifyBeginMethodOnVertxVersion3() {
+		try {
+			final Method method = SqlConnection.class.getMethod( "begin" );
+			final Class<?> returnType = method.getReturnType();
+			// if it returns Transaction
+			return Transaction.class.equals( returnType )
+					// then we're on Vert.x v. 3.9
+					? method
+					// else it's Vert.x 4
+					: null;
+		}
+		catch (NoSuchMethodException e) {
+			return null; //some new Vert.x version? No need for Vertx 3 compatibility in this case.
+		}
 	}
 
 	@Override
 	public CompletionStage<Void> beginTransaction() {
-		transaction = connection.begin();
-		return voidFuture();
-//		return execute("begin");
+		CompletionStage<Transaction> transactionStage = RUNNING_ON_VERTX3
+				? CompletionStages.completedFuture( beginTransactionVertx3() )
+				: Handlers.<Transaction>toCompletionStage( connection::begin );
+		return transactionStage.thenAccept( tx -> transaction = tx );
+	}
+
+	private Transaction beginTransactionVertx3() {
+		try {
+			return (Transaction) BEGIN_ON_3.invoke( connection );
+		}
+		catch (IllegalAccessException | InvocationTargetException e) {
+			throw new AssertionFailure( "Error beginning transaction on Vert.x 3 compatibility mode. Probably best to upgrade Vert.x to version 4", e );
+		}
 	}
 
 	@Override
 	public CompletionStage<Void> commitTransaction() {
-		return Handlers.toCompletionStage(
-				handler -> transaction.commit(
-						ar -> {
-							transaction = null;
-							handler.handle( ar );
-						}
-				)
-		);
-//		return execute("commit");
+		return Handlers.<Void>toCompletionStage(transaction::commit)
+				.whenComplete( (v, x) -> transaction = null );
 	}
 
 	@Override
 	public CompletionStage<Void> rollbackTransaction() {
-		return Handlers.toCompletionStage(
-				handler -> transaction.rollback(
-						ar -> {
-							transaction = null;
-							handler.handle( ar );
-						}
-				)
-		);
-//		return execute("rollback");
+		return Handlers.<Void>toCompletionStage(transaction::rollback)
+				.whenComplete( (v, x) -> transaction = null );
 	}
 
 	@Override
 	public void close() {
-		connection.close();
+		if (RUNNING_ON_VERTX3) {
+			try {
+				CLOSE_ON_3.invoke( connection );
+			}
+			catch (IllegalAccessException | InvocationTargetException e) {
+				throw new AssertionFailure( "Error closing connection on Vert.x 3 compatibility mode. Probably best to upgrade Vert.x to version 4", e );
```

Comment:
Is there any likelihood that upgrading to Vert.x 4 is really a solution to whatever problem is causing this failure? I think this error message is most likely unhelpful.

Reply:
Shouldn't these exceptions only occurs if the `close()` method doesn't exist or is not accessible?

If it doesn't exist upgrading should solve the issue.
If the problem is an IllegalAccess I'm not sure the fact that we also show the cause should help.

Do you have any suggestions about what to add?
Would something like this be better?
```
SqlClientConnection#close() method not found or not accessible. Check you are using a supported Vert.x version (3 or 4)
```

Reply:
I don't think they can possibly occur if `close()` doesn't exist, since we would not have been able to obtain the method ref.

They can only occur if the `close()` methods throw, it seems to me.